File: src/blocks/types/dirt.go
use items;
use render;

// DirtBlock is a cubeeee of dirttttt
struct DirtBlock {
    Block
    lifespan, health f32 
}

// newDirtBlock creates and returns a new DirtBlock
fn newDirtBlock() &DirtBlock {
	return &DirtBlock {
        lifespan: 15,
        health: 15,
    }
}

// hit is called with the DirtBlock is hit
fn hit(fn hit(&self, with interface{}, power f32)self, with interface{}, power f32) []items.Item {
    return nil
}

// kill is called when the DirtBlock is obliterated. Returns
// Dirt.
// items that might be dropped with the killable dies.
fn kill(&self) []items.Item {
    b.health = 0
    return nil //[]items.Item{items.Dirt}
}

// IsAlive returns true if the DirtBlock is still intact
fn IsAlive(&self) bool {
    return b.health > 0
}

// HealthLeft returns the number of health points left on
// the DirtBlock
fn HealthLeft(&self) f32 {
    return b.health
}

// Lifespan returns the lifespan of health points on the
// DirtBlock
fn Lifespan(&self) f32 {
    return b.lifespan
}

// render renders the DirtBlock
fn render(&self, c &render.Camera) {

}

File: src/blocks/types/dirt.rs
use items;
use render;

/// DirtBlock is a cubeeee of dirttttt
struct DirtBlock {
    lifespan: f32,
    health: f32,
}

impl DirtBlock {
    /// Creates and returns a new DirtBlock
    fn new() -> Self {
        return Self {
            lifespan: 15,
            health: 15,
        }
    }
}

impl Block for DirtBlock {
    /// Called with the DirtBlock is hit
    fn hit(&self, with: impl Any, power: f32) -> &[items.Item] {
        &[]
    }

    /// Called when the DirtBlock is obliterated. Returns
    /// Dirt.
    fn kill(&self) -> &[items.Item] {
        self.health = 0;
        &[]
    }

    // is_alive returns true if the DirtBlock is still intact
    fn is_alive(&self) -> bool {
        self.health > 0
    }

    // health_left returns the number of health points left on
    // the DirtBlock
    fn health_left(&self) -> f32 {
        self.health
    }

    // lifespan returns the lifespan of health points on the
    // DirtBlock
    fn lifespan(&self) -> f32 {
        self.lifespan
    }

    // render renders the DirtBlock
    fn render(&self, c: &render::Camera) {

    }
}
